source/adapters/hip/memory.cpp

```diff
@@ -243,9 +265,31 @@ UR_APIEXPORT ur_result_t UR_APICALL urMemGetInfo(ur_mem_handle_t hMemory,
   switch (MemInfoType) {
   case UR_MEM_INFO_SIZE: {
     try {
-      size_t AllocSize = 0;
-      UR_CHECK_ERROR(hipMemGetAddressRange(
-          nullptr, &AllocSize, std::get<BufferMem>(hMemory->Mem).Ptr));
+      const auto MemVisitor = [](auto &&Mem) -> size_t {
+        using T = std::decay_t<decltype(Mem)>;
```

Comment:
Why is this templated? Surely we know the type here?

If it's mean

Reply:
Whoops. Sent that without finishing my sentence.

Just wondering why we need std::visit and a generic lambda.

Reply:
Not necessarily - It can be a struct with the call operator overloaded for each of the types of the variant.
I've seen before people define a helper struct to which you can add the overloads:

i.e. 
```cpp
template<class... Ts> struct Overload : Ts...{
using Ts::operator()...;
};
template<class... Ts> Overload(Ts...) -> Overload<Ts...>;

// then you can something like this.
std::variant<int, float, long> var;
std::visit(Overload{
[](int x){ /* run on int*/},
[](float x){ /* run on float*/},
[](long x){ /* run on long*/},
}, var);

```